### PR TITLE
extract serde crate

### PR DESCRIPTION
**Pull request by kevinheavey**
#### Problem
`solana_sdk::deserialize_utils` imposes a `solana_sdk` dep on `solana_client`

#### Summary of Changes
Move to its own crate and re-export with deprecation notice

### CODE REVIEW

**Comment by febo** [comment]
Wonder if this helper should be added to `solana-bincode` crate, since the comments suggest that it is bincode specific – although there is not a "direct" dependency to it.

cc: @joncinque

**Comment by joncinque** [comment]
> Wonder if this helper should be added to solana-bincode crate, since the comments suggest that it is bincode specific – although there is not a "direct" dependency to it.

Looking through the usage of `default_on_eof`, it's true that it's exclusively targeted at `bincode` deserialization. On the flipside, the code is dependent on matching a specific string version of an `std::io::Error` (https://doc.rust-lang.org/src/std/io/error.rs.html#472), so there's an argument that it's more general than `bincode` and *could* maybe be used outside of it.

On the other-other hand, `default_on_eof` is always used independently of what currently exists in `solana-bincode`, so we might be best served with a new `solana-serde` crate. It's not a strong preference, but that's where I lean.

**Comment by kevinheavey** [comment]
I have no strong opinion either way

**Comment by febo** [comment]
A `solana-serde` feels a bit more generic and allows for other "serde" helpers to be easily added without creating a new crate – I am leaning towards renaming this crate to `solana-serde`.

**Review by joncinque — DISMISSED** [review]
Looks good to me! @yihau can you accept ownership of `solana-serde`?

**Review by febo — DISMISSED** [review]

**Comment by yihau** [comment]
✅

**Review by febo — DISMISSED** [review]

File: sdk/src/lib.rs [hunk 1]
```diff
@@ -191,6 +190,8 @@ pub use solana_sdk_macro::declare_id;
 pub use solana_sdk_macro::pubkeys;
 #[deprecated(since = "2.1.0", note = "Use `solana-secp256k1-recover` crate instead")]
 pub use solana_secp256k1_recover as secp256k1_recover;
+#[deprecated(since = "2.1.0", note = "Use `solana-serde` crate instead")]
```
Comment (febo) [inline sdk/src/lib.rs]: Should this say `2.2.0` instead?

Comment (kevinheavey) [reply sdk/src/lib.rs]: yes good catch

**Review by febo — APPROVED** [review]
Looks good to me.

**Comment by kevinheavey** [comment]
@febo could you merge please? Looks like automerge isn't feeling it today

**Pull request merged**